File: solutions_1484496_0/java/Phicar/equal.java
import java.util.*;
import java.io.*;
public class equal{
public static void main(String args[]) throws IOException{
BufferedReader lector = new BufferedReader(new InputStreamReader(System.in));
int tod = Integer.parseInt(lector.readLine());
//HashMap<Long,Vector<Integer>> tal = new HashMap<Long,Vector<Integer>>();
for(int i = 0;i<tod;i++){
HashMap<Long,Vector<Integer>> tal = new HashMap<Long,Vector<Integer>>();
System.out.println("Case #"+(i+1)+":");
String jod = lector.readLine();
String tmp[] = jod.split(" ");
//System.out.println(jod);
boolean sicas = false;
for(int n = 1;n<(1<<(tmp.length-1));n++){
long sum = 0;
Vector<Integer> tm = new Vector<Integer>();
for(int m = 0;m<(tmp.length-1);m++){
if(((n>>m)&1)==1){
tm.add(Integer.parseInt(tmp[m+1]));
sum+=Long.parseLong(tmp[m+1]);
}
}
//System.out.println(tm);
if(sum>0 && tal.get(sum)!=null){
imprimir(tal.get(sum));
imprimir(tm);
sicas = true;
break;
}else
tal.put(sum,tm);
}
if(!sicas)System.out.println("Impossible");
}
}
public static void imprimir(Vector<Integer> a){
Collections.sort(a);
for(int n = 0;n<a.size();n++)
System.out.print(a.get(n)+((n==a.size()-1)?"":" "));
System.out.println();
} 
}

File: solutions_1484496_0/Python/ronnodas/equal.py
T=int(input())
for t in range(T):
    N = 2**20
    print('Case #',t+1,':',sep = '')
    ns = [int(x) for x in input().split()][1:]
    sums = {}
    ns.sort()
    for i in range(1,N):
        s = 0
        for d in range(20):
            if i&(1<<d):
                s+= ns[d]
        if s in sums:
            j = sums[s]
            for k in [j,i]:
                ls = []
                for d in range(20):
                    if k&(1<<d):
                        ls.append(ns[d])
                print(*ls)
            break
        else:
            sums[s] = i
    else:
        print("Impossible")
